app/view/form/LayerTreeFilter.js

```diff
@@ -0,0 +1,109 @@
+/**
+ * Panel with filtering tools for the layer tree.
+ *
+ * @class CpsiMapview.view.from.LayerTreeFilter
+ */
+Ext.define('CpsiMapview.view.form.LayerTreeFilter', {
+    extend: 'Ext.form.FieldSet',
+
+    xtype: 'cmv_layertreefilter',
+
+    requires: [
+        'CpsiMapview.controller.form.LayerTreeFilter'
+    ],
+
+    controller: 'cmv_layertreefilter',
+
+    config: {
+        /**
+         * If baselayers shall be filtered by text filter.
+         */
+        doFilterBaseLayers: true
+    },
+
+    /**
+     * The ID of the filter created by the text field.
+     */
+    TEXT_FILTER_ID: 'cpsi-tree-layer-text-filter',
+
+    /**
+     * The ID of the filter created by the checkbox.
+     */
+    VISIBLE_LAYER_FILTER_ID: 'cpsi-tree-visible-layer-filter',
+
+    /**
+     * The ID of the filter to hide empty groups.
+     */
+    HIDE_EMPTY_GROUPS_FILTER_ID: 'cpsi-tree-hide-empty-groups',
+
+    items: [{
+        xtype: 'textfield',
+        emptyText: 'Search for layers',
+        hideTrigger: true,
+        anchor: '100%',
+        maxWidth: 250,
+        triggers: {
+            clearTreeFilterText: {
+                cls: 'x-form-clear-trigger',
+                handler: 'clearText'
+            }
+        },
+        listeners: {
+            change: 'onSearch'
+        }
+    },
+    {
+        xtype: 'checkboxfield',
+        boxLabel: 'Show Visible',
+        listeners: {
+            change: 'filterVisibleLayers'
+        }
+    }],
+
+    initComponent: function() {
+        var me = this;
+        me.callParent(arguments);
+
+        // we need to set a filter that hides groups that do not have visible children
+        var emptyGroupFilter = Ext.util.Filter({
+            id: me.HIDE_EMPTY_GROUPS_FILTER_ID,
+            filterFn: function (node) {
+                if (node.hasChildNodes()) {
+                    var keep = false;
+                    node.cascade(function (cascadeNode) {
+                        // search for any descending node that is checked
+                        if (!cascadeNode.hasChildNodes() && cascadeNode.get('checked')) {
```

Comment:
Just to confirm the expected behaviour here is that if layers are switched on then their parent folders will be displayed for any searches?

So for example the Road Schedule layer is turned on (in the "Road Network Layers" folder) and then a search is made which does not match any layers in this folder, the folder remains visible (and cannot be expanded):

Step 1:

![image](https://user-images.githubusercontent.com/490840/183504422-e112f82c-1e0a-4e41-b8f5-7d6bbe89d4a5.png)

Step 2:

![image](https://user-images.githubusercontent.com/490840/183504829-9210db0e-4ea0-4187-9778-1a1d2e7c1d54.png)

Would it make sense to hide empty folders when searching, even if they contain a visible layer?

Reply:
@geographika yes true. your described behaviour is correct.

Good question, this could indeed useful.  I will check how much effort that is.